### DISCUSSION TITLE
menu level

### DISCUSSION THREAD

**Discussion by ODA669** [OrchardCMS/OrchardCore#13923] [General]
I want to insert a span in the part of the dropdown-item but I don't want it to be duplicated for each item if not before and I can't do it it duplicates when I try to insert it and besides that I want it to show the name of the dropdown menu in That span, if anyone knows how to manipulate the menu levels or how I can add it without duplicating it, it would help me a lot, thanks.

    int level = (int)Model.Level;
    TagBuilder tag = Tag(Model, "li");
    TagBuilder parentTag = null;

    if ((bool)Model.HasItems)
    {
        tag.AddCssClass("dropdown ");
        parentTag = Tag(Model, "ul");
        parentTag.AddCssClass("dropdown-menu dropdown-content slide-left");
        parentTag.Attributes["id"] = "expanded";
    }

    Model.Metadata.Alternates.Clear();
    Model.Classes.Clear();
    Model.Metadata.Type = "MenuItemLink";

    @if (Model.Level > 0)
    {
        Model.Classes.Add("dropdown-item");
         //// i try to put here but it duplicated in each element
        //<span style="color:white">@Model.Items</span>
    }
    else
    {
        Model.Classes.Add("nav-link dropdown-toggle");
       
     
        if (Model.HasItems)
        {
  
            Model.Classes.Add("");

            Model.Attributes.Add("data-bs-toggle", "dropdown-toggle");
        }
    }
    tag.InnerHtml.AppendHtml(await DisplayAsync(Model));
    if (parentTag != null)
    {
        
        foreach (var item in Model.Items)
        {
            item.Level = level + 1;
            item.ParentTag = parentTag;
            parentTag.InnerHtml.AppendHtml(await DisplayAsync(item));
         
        }
        tag.InnerHtml.AppendHtml(parentTag);
    }
}
@tag

**Comment by jtkech** [comment]
I assume that this is when `.HasItems` is true that the current item is a parent having childs, can be checked further with `parentTag` being not null and where all childs are rendered, notice that the `.Level` of childs is set to the parent value + 1. Then when rendering a child the same shape view may be used, and a child item may also be a parent having childs.

**Comment by ODA669** [comment]
I have done this verification but it is still shown on each dopdown-item and if I place it outside the if it shows it at level 0 I want a single span to appear before showing the dropdowns,its juste a title with the name to button menu in livel 0


bool isLevel1Displayed = false;

if (Model.Level > 0)
        {
            Model.Classes.Add("dropdown-item");

            if (Model.Level == 1 && !isLevel1Displayed)
            {
                <span style="color:white">Texto antes de los valores de nivel 1: </span>
                isLevel1Displayed = true;
            }
        }

**Reply by jtkech** [reply]
So as described, try to put your code for example in the `if ((bool)Model.HasItems)` block.

**Reply by jtkech** [reply]
Can you provide a simpe image describing what you want to render, so that I could give it a try when I will have time ;)

**Comment by ODA669** [comment]
dont work

**Reply by jtkech** [reply]
Can you provide a simpe image describing what you want to render, so that I could give it a try when I will have time ;)

**Comment by ODA669** [comment]
![Sans titre](https://github.com/OrchardCMS/OrchardCore/assets/60682935/115d8e92-997d-4f98-93fa-67be1da521f3)
in the line Texto antes de los valores de nivel 1: i want to render  the name of button ACCUEIL only one time in the first place

**Reply by jtkech** [reply]
Hmm, so ACCUEIL could be a parent menu item, so that when you click it you get the 2 following child items, right?

**Comment by ODA669** [comment]
yes

**Reply by jtkech** [reply]
Okay I will try something and share the result

**Reply by ODA669** [reply]
thanks

**Reply by jtkech** [reply]
Is the following okay? Are you French?

![Menu](https://github.com/OrchardCMS/OrchardCore/assets/8586360/2ccf47a3-e4f8-4d13-91b9-6ca4c59d8d8b)

**Reply by ODA669** [reply]
it's the dsame resul that i have

**Reply by jtkech** [reply]
So what's the problem?

**Reply by jtkech** [reply]
Hmm, maybe you only want the Span Text be rendered once, only before the first child item?

**Reply by ODA669** [reply]
i need to show  the name of button in the first span only, dont show in the second span

**Reply by jtkech** [reply]
Ah okay, let me try something

**Reply by ODA669** [reply]
but it ok,. thanks

**Reply by jtkech** [reply]
Is it okay?

![Menu](https://github.com/OrchardCMS/OrchardCore/assets/8586360/02562b9b-2b53-4199-99bc-2f8173c26882)

**Reply by ODA669** [reply]
yes how this but it  can take the name of button , because in my menu there are many buttons with the same behavor each dropdown with the name of diferent button

**Reply by jtkech** [reply]
If so, here the part of the code managing it


    @if ((bool?)Model.FirstItem == true)
    {
        <span style="color:white">Span Text Before</span>
    }

    tag.InnerHtml.AppendHtml(await DisplayAsync(Model));

    if (parentTag != null)
    {
        var firstItem = true;
        foreach (var item in Model.Items)
        {
            if (firstItem)
            {
                item.FirstItem = true;
                firstItem = false;
            }

            item.Level = level + 1;
            item.ParentTag = parentTag;
            parentTag.InnerHtml.AppendHtml(await DisplayAsync(item));
        }

        tag.InnerHtml.AppendHtml(parentTag);
    }

**Reply by jtkech** [reply]
> yes how this but it can take the name of button

Where this button and name is defined?

**Reply by ODA669** [reply]
thanks so much!!!
i will try it

**Comment by ODA669** [comment]
hi i have a other question

 ""<span style="color:white">Span Text Before</span>""

for show in this part " Span Text Before " ACCUEIL that is the name of button to menu  i can't find how to show this, i try with  Model.ContentItem.Content 
Model.ContentItem.Content .Items
Model.Items
Model.ContentItem
Model.Menu

but none work, do you can help me?

**Reply by jtkech** [reply]
Maybe, try the following.

    @if ((bool?)Model.FirstItem == true)
    {
        <span style="color:white">@Model.ParentName</span>
    }

    tag.InnerHtml.AppendHtml(await DisplayAsync(Model));

    if (parentTag != null)
    {
        var firstItem = true;
        foreach (var item in Model.Items)
        {
            if (firstItem)
            {
                item.FirstItem = true;
                item.ParentName = Model.ContentItem.DisplayText;
                firstItem = false;
            }

            item.Level = level + 1;
            item.ParentTag = parentTag;
            parentTag.InnerHtml.AppendHtml(await DisplayAsync(item));
        }

        tag.InnerHtml.AppendHtml(parentTag);
    }

**Reply by ODA669** [reply]
Yes works,, thanks so much

**Reply by jtkech** [reply]
It works on my side, in the span above the first child item, it renders the name of the parent item, both parent and child items being `Link Menu Item`, but I may have misunderstood the goal.

**Reply by jtkech** [reply]
Ah okay, just saw that it also works on your side ;)
